presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java

```diff
@@ -701,13 +701,26 @@ public PlanNode visitValues(ValuesNode node, RewriteContext<Set<Symbol>> context
         @Override
         public PlanNode visitApply(ApplyNode node, RewriteContext<Set<Symbol>> context)
         {
-            ImmutableSet<Symbol> inputContext = ImmutableSet.<Symbol>builder()
+            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
+
+            // prune not used correlation symbols
+            Set<Symbol> subquerySymbols = DependencyExtractor.extractUnique(subquery);
+            List<Symbol> newCorrelation = node.getCorrelation().stream()
+                    .filter(subquerySymbols::contains)
+                    .collect(toImmutableList());
+
+            Set<Symbol> inputContext = ImmutableSet.<Symbol>builder()
                     .addAll(context.get())
-                    .addAll(node.getCorrelation())
+                    .addAll(newCorrelation)
                     .build();
             PlanNode input = context.rewrite(node.getInput(), inputContext);
-            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
-            return new ApplyNode(node.getId(), input, subquery, node.getCorrelation());
+
+            if (!inputContext.stream().anyMatch(subquery.getOutputSymbols()::contains)) {
+                // prune ApplyNode as subquery symbols are not used
```

Comment:
Eh... I really don't like when you are right ;)

It is much more complicated than I previously thought. `Apply` node can be only pruned when it is placed in projection (or in `VALUES`) and its output is not used. To do that I need to remove `Apply` when I prune matching `Expression`. In this place I have no context what introduced given `Apply` node so I cannot be removed.

I am going to rework this.